src/facility.py

```diff
@@ -0,0 +1,311 @@
+        suffix = '",oauth_signature_method="PLAINTEXT",oauth_version="1.0"'
+        if self.access_token is None and self.access_secret is None:
+            return f'{prefix}{suffix}'
+        ampersand = "%26"
+        return f'{prefix}{ampersand}{self.access_secret}",oauth_token="{self.access_token}{suffix}'
+
+
+class ClientBase:
+    """
+    Base class for HTTP client classes.
+    """
+    def __init__(self,
+                 base_uri: str, *,
+                 headers: Optional[Dict[str, str]] = None,
+                 oauth: Optional[OAuthSettings] = None,
+                 max_retries: int = 0):
+        self.base_uri = base_uri.rstrip("/")
+        self.headers = headers
+        self.oauth = oauth
+        self.max_retries = int(max_retries)
+        self.session = None
+
+    def send_request(self, method: str, uri: str, *,
+                     query: Optional[dict] = None,
+                     request: Any = None,
+                     headers: Optional[Dict[str, str]] = None) -> requests.Response:
+        headers_ = dict()
+        if self.oauth is not None:
+            headers_['Authorization'] = self.oauth.to_header()
+        if request is not None:
+            headers_['Content-Type'] = 'application/json'
+        if self.headers is not None:
+            headers_.update(self.headers)
+        if headers is not None:
+            headers_.update(headers)
+        if self.session is None:
+            self.session = requests.sessions.Session()
+            if self.max_retries > 0:
+                self.session.mount("http://", requests.adapters.HTTPAdapter(max_retries=self.max_retries))
+                self.session.mount("https://", requests.adapters.HTTPAdapter(max_retries=self.max_retries))
+        resp = self.session.request(
+            method=method,
+            url=f"{self.base_uri}{uri}",
+            params=query or dict(),
+            json=request,
+            headers=headers_
+        )
+        try:
+            _ = resp.content  # Consume socket so it can be released
+        except (ChunkedEncodingError, ContentDecodingError, RuntimeError):
+            resp.raw.read(decode_content=False)
+        return resp
+
+    def close(self):
+        if self.session is not None:
+            self.session.close()
+            self.session = None
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, *args):
+        self.close()
+
+
+class Enum(enum.Enum):
+
+    @classmethod
+    def get(cls, value: str, default: Optional["Enum"] = None) -> Optional["Enum"]:
+        return cls.get_value_map().get(value, default)
+
+    @classmethod
+    def get_value_map(cls):
+        key = '_value_map'
+        v2e = getattr(cls, key, None)
+        if v2e is None:
+            v2e = {s.value: s for s in cls}
+            setattr(cls, key, v2e)
+        return v2e
+
+
+def string_to_bool(value: Union[str, bool, type(None)]) -> Optional[bool]:
```

Comment:
I'm surprised this function allows so many types of input.

Reply:
I guess it's the Python way.